In [ ]:
#Load Liabraries
import numpy as np
import pandas as pd

In [ ]:
book = pd.read_csv('/content/book.csv',encoding='latin-1') #Read Data

In [ ]:
book

Unnamed: 0  ...  Book.Rating
0              1  ...            5
1              2  ...            3
2              3  ...            6
3              4  ...            8
4              5  ...            6
...          ...  ...          ...
9995        9996  ...            7
9996        9997  ...            9
9997        9998  ...            7
9998        9999  ...            8
9999       10000  ...            6

[10000 rows x 4 columns]

In [ ]:
book.describe()

Drop Column

In [ ]:
book = book.drop(['Unnamed: 0'],axis = 1)

In [ ]:
book

User.ID                                         Book.Title  Book.Rating
0      276726                                Classical Mythology            5
1      276729                                       Clara Callan            3
2      276729                               Decision in Normandy            6
3      276736  Flu: The Story of the Great Influenza Pandemic...            8
4      276737                             The Mummies of Urumchi            6
...       ...                                                ...          ...
9995   162121       American Fried: Adventures of a Happy Eater.            7
9996   162121                              Cannibal In Manhattan            9
9997   162121                    How to Flirt: A Practical Guide            7
9998   162121                                           Twilight            8
9999   162129                       Kids Say the Darndest Things            6

[10000 rows x 3 columns]

In [ ]:
book.columns=['userID','title','rating']

In [ ]:
len(book.userID.unique())

2182

In [ ]:
len(book.title.unique())


9659

Duplicate

In [ ]:
book.userID.duplicated()


0       False
1       False
2        True
3       False
4       False
        ...  
9995     True
9996     True
9997     True
9998     True
9999    False
Name: userID, Length: 10000, dtype: bool

In [ ]:
book.userID.duplicated().sum()

7818

In [ ]:
book.loc[book.userID.duplicated(),:]

userID                                              title  rating
2     276729                               Decision in Normandy       6
8     276747  Under the Black Flag: The Romance and the Real...       9
9     276747            Where You'll Find Me: And Other Stories       8
10    276747                        Nights Below Station Street       7
11    276747  Hitler's Secret Bankers: The Myth of Swiss Neu...       7
...      ...                                                ...     ...
9994  162121  I Should Have Stayed Home: The Worst Trips of ...       8
9995  162121       American Fried: Adventures of a Happy Eater.       7
9996  162121                              Cannibal In Manhattan       9
9997  162121                    How to Flirt: A Practical Guide       7
9998  162121                                           Twilight       8

[7818 rows x 3 columns]

In [ ]:
book.duplicated().sum()


2

In [ ]:
book.loc[book.duplicated(),:]


userID                      title  rating
5051    2152  Le nouveau soleil de Teur       7
7439    3757        The Magician's Tale       7

# Creating a Pivot table

In [ ]:

user_df=book.pivot_table(index='userID',columns='title',values='rating').reset_index(drop=True)
userID=user_df.index
title=user_df.columns
user_df

title   Jason, Madison &amp  ...  Ã?Â?bermorgen.
0                       NaN  ...             NaN
1                       NaN  ...             NaN
2                       NaN  ...             NaN
3                       NaN  ...             NaN
4                       NaN  ...             NaN
...                     ...  ...             ...
2177                    NaN  ...             NaN
2178                    NaN  ...             NaN
2179                    NaN  ...             NaN
2180                    NaN  ...             NaN
2181                    NaN  ...             NaN

[2182 rows x 9659 columns]

# Creating index from UserID.

In [ ]:

user_df.index=book.userID.unique()

#Filling the NA values with zero


In [ ]:
user_df.fillna(0, inplace =True)
user_df

title    Jason, Madison &amp  ...  Ã?Â?bermorgen.
276726                   0.0  ...             0.0
276729                   0.0  ...             0.0
276736                   0.0  ...             0.0
276737                   0.0  ...             0.0
276744                   0.0  ...             0.0
...                      ...  ...             ...
162107                   0.0  ...             0.0
162109                   0.0  ...             0.0
162113                   0.0  ...             0.0
162121                   0.0  ...             0.0
162129                   0.0  ...             0.0

[2182 rows x 9659 columns]

In [ ]:

# Import libraries to find similarities 
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine,correlation

In [ ]:
# FInding Similarities
user_sim=1-pairwise_distances(user_df,metric='cosine')
user_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [ ]:

user_sim_df=pd.DataFrame(user_sim)
user_sim_df.index=book.userID.unique()
user_sim_df.columns=book.userID.unique()
user_sim_df

276726  276729  276736  276737  ...  162109  162113  162121  162129
276726     1.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
276729     0.0     1.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
276736     0.0     0.0     1.0     0.0  ...     0.0     0.0     0.0     0.0
276737     0.0     0.0     0.0     1.0  ...     0.0     0.0     0.0     0.0
276744     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
...        ...     ...     ...     ...  ...     ...     ...     ...     ...
162107     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
162109     0.0     0.0     0.0     0.0  ...     1.0     0.0     0.0     0.0
162113     0.0     0.0     0.0     0.0  ...     0.0     1.0     0.0     0.0
162121     0.0     0.0     0.0     0.0  ...     0.0     0.0     1.0     0.0
162129     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     1.0

[2182 rows x 2182 columns]

In [ ]:
# Filling the diagonal cells with zero 
np.fill_diagonal(user_sim,0)
user_sim_df

276726  276729  276736  276737  ...  162109  162113  162121  162129
276726     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
276729     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
276736     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
276737     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
276744     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
...        ...     ...     ...     ...  ...     ...     ...     ...     ...
162107     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
162109     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
162113     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
162121     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
162129     0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0

[2182 rows x 2182 columns]

# Most similar users

In [ ]:

user_sim_df.idxmax(axis=1)[0:20]

276726    276726
276729    276726
276736    276726
276737    276726
276744    276726
276745    276726
276747    276726
276748    161677
276751    276726
276754    276726
276755    276726
276760    276726
276762    276726
276768    276726
276772      1491
276774    278543
276780    276726
276786    276726
276788    276726
276796    276726
dtype: int64

# Similarity between userID 276726 and 276729


In [ ]:
book[(book['userID']==276726) | (book['userID']==276729)]

userID                 title  rating
0  276726   Classical Mythology       5
1  276729          Clara Callan       3
2  276729  Decision in Normandy       6

In [ ]:
count_of_rating=pd.DataFrame(book.groupby('title')['rating'].count().sort_values(ascending=False))
count_of_rating.head()

rating
title                                                  
Fahrenheit 451                                        5
Vanished                                              4
Charlie and the Chocolate Factory                     4
Stardust                                              4
The Amber Spyglass (His Dark Materials, Book 3)       4

In [ ]:

avg_rating=pd.DataFrame(book.groupby('title')['rating'].mean().sort_values(ascending=False))
avg_rating

rating
title                                                     
Clear Light of Day                                    10.0
Substitute Teaching: A Handbook for Hassle-Free...    10.0
Daughter of Camelot (Merlin's Legacy, No 6)           10.0
The Resourceful English Teacher                       10.0
Das verbotene Lied.                                   10.0
...                                                    ...
Time and Chance                                        1.0
Upsetting the Balance (Worldwar Series, Volume 3)      1.0
Having Our Say: The Delany Sisters' First 100 Y...     1.0
Saucer                                                 1.0
THIS SIDE OF PARADISE (A Scribner Classic)             1.0

[9659 rows x 1 columns]

In [ ]:
sum_of_rating=pd.DataFrame(book.groupby('title')['rating'].sum().sort_values(ascending=False))
sum_of_rating

rating
title                                                  
Fahrenheit 451                                       35
The Subtle Knife (His Dark Materials, Book 2)        35
Stardust                                             34
The Amber Spyglass (His Dark Materials, Book 3)      33
The Face                                             27
...                                                 ...
At Wits End                                           1
The Honk and Holler Opening Soon                      1
THIS SIDE OF PARADISE (A Scribner Classic)            1
The Door into Summer                                  1
The Selfish Gene                                      1

[9659 rows x 1 columns]

**This is the books Recommended by rankings**


#**Fahrenheit 451**
#**The Subtle Knife (His Dark Materials, Book 2)**	
#**Stardust**
#**The Amber Spyglass (His Dark Materials, Book 3)**	
#**The Face**


